### ISSUE TITLE
Build Database as SWIM runs

### ISSUE THREAD

**Issue by bettinardi** [tlumip/tlumip#52]
It would be ideal (but low priority) for SWIM to write data to the VizDB by year, or better yet by module.  It would be best if this happened in Parallel as opposed to in series, meaning that after a year or module finished it start to move the data to the database, but also allowed the next process in the SWIM stream to start.

We should think about ways that this could be coded.

**Comment by gregmacfarlane** [comment]
It would be easy to do this in Linux with the `nohup` command. We'll have to see if Windows has anything similar.

**Comment by bettinardi** [comment]
Copying notes from issue 101 to close issue 101:

We should also consider having TSTEPs functionality that both indicates which years to run DB creator on and which years to delete individual year DBs on. I'm thinking of this, because as we push the base run year into the future, we might want to keep database summaries from "historic" years in the repo (in the outputs for previous historic years) and not have the process delete them. So we should be able to specify in TSTEPS what years we want the database run for and what years should have the year specific databases deleted.

**Comment by bettinardi** [comment]
WOC 7 summary - 

This task adds a new feature to tsteps.  It allows the user easier ability to configure how the SWIM databases build.  At the end of each year in tsteps there should be a DB (or similar) field.  We might need something besides "1".  The idea would be that non-end years would get a different flag to start off and start building the DB for that year without running in series; the next year could start without waiting for the DB to finish building.  The last year would get a "1".  The "Viz" step would be changed to just compile existing datasets.

We would want to rethink the deletion of the year specific databases.  Currently the code does it automatically.  I think if a year is specified to be created, the first step could be to delete existing databases.  For the end, I think we want to add a DB_delete (or similar) feature to tsteps to indicate if the year specific DBs should be deleted or not.

There are cases where a user may want to run SWIM for historic years and save year DBs, then copy those over to future years to build Databases with both historic and future years.  There should probably be a scoping step up front for this issue to clarify and design the exact end state of the implementation before coding starts.

**Comment by bstabler** [comment]
- update tsteps to:
  - if user puts a 1 in the year row, viz column then if not last year, create db and don't wait
  - if user puts a 1 in the last year row, viz column then create db and then compile all year dbs found, and then delete all dbs with a 1 in their year row, viz column

**Comment by goreaditya** [comment]
The following modifications were made to address this issue:
- A new property token was created `@VIZ.FINAL.YEAR@` that is used to keep track of the last year the VIZ module will be called based on *tsteps.csv*. This token will be used during the building of model setup.
- A new property was created `viz.final.year` in the *globalTemplate.properties* file. This property will be used by the VIZ java module to combine all the databases after building the database for final year.
- The *build_run.py* script was updated to create a run script that calls database builder for years marked as 1 in the VIZ column in *tsteps.csv*. The run script is written such the database builder calls for all years are initiated in a separate independent command shell except for the final year.
- R scripts were updated to address the issue of package updates made to `RSQLite`.

These modifications were tested using following *tsteps.csv* configuration and ran successfully.
| Year | SI   | NED  | ALD  | POPSIMSPG1 | AA   | POPSIMSPG2 | PT   | CT   | TA   | TR   | PT - Logsums Only | SL - Generate Select Link Data | SL - Append Select Link To Trips | VIZ  | MICROVIZ | SWIMR_reference | SWIMR_compare | SWIMR_PopEmp |
| ---- | ---- | ---- | ---- | ---------- | ---- | ---------- | ---- | ---- | ---- | ---- | ----------------- | ------------------------------ | -------------------------------- | ---- | -------- | --------------- | ------------- | ------------ |
| 18   |      |      |      |            | i    |            |      |      |      |      |                   |                                |                                  |      |          |                 |               |              |
| 19   | 1    | 1    | i    | i          | 1    | i          | i    | i    | i    | i    | i                 |                                |                                  |      |          |                 |               |              |
| 20   | 1    | 1    | 1    | 1          | 1    | 1          | 1    | 1    | 1    | 1    | 1                 |                                |                                  | 1    |          |                 |               |              |
| 21   | 1    | 1    | 1    | 1          | 1    |            |      |      |      |      |                   |                                |                                  |      |          |                 |               |              |
| 22   | 1    | 1    | 1    | 1          | 1    |            |      |      |      |      |                   |                                |                                  |      |          |                 |               |              |
| 23   | 1    | 1    | 1    | 1          | 1    | 1          | 1    | 1    | 1    | 1    | 1                 |                                |                                  | 1    |          |                 |               |              |
| 24   | 1    | 1    | 1    | 1          | 1    |            |      |      |      |      |                   |                                |                                  |      |          |                 |               |              |
| 25   | 1    | 1    | 1    | 1          | 1    |            |      |      |      |      |                   |                                |                                  |      |          |                 |               |              |
| 26   | 1    | 1    | 1    | 1          | 1    | 1          | 1    | 1    | 1    | 1    | 1                 |                                |                                  | 1    |          |                 |               |              |

**Comment by goreaditya** [comment]
The _build_run.py_ and _module_commands.py_ were modified so that if VIZ crashes then it will create a file called `viz_failed.txt` and document the error code. This error code (if exist) will be read just before next VIZ run and the model run is stopped if the code is different then 0. The code has been tested by inserting a `stop()` code in the _build_VIZ_DB.R_ script and allowing the VIZ to crash.

**Comment by bettinardi** [comment]
Just a note for future work that this feature does now work as spec'd - it runs along side models year runs to create a faster database build at the end (where the majority of years are already completed).  However there are a number of ways that this is problematic and prone to user error.  If the user forgets to generate databases year by year, it's very difficult currently to go back and generate a database outside of a run.  This is an area that will likely need further improvement.
